tempesta_fw/sock.c

```diff
@@ -1366,7 +1375,8 @@ ss_tx_action(void)
 			break;
 		case SS_CLOSE:
 			if (!((1 << sk->sk_state)
-			      & (TCPF_ESTABLISHED | TCPF_SYN_SENT)))
+			      & (TCPF_ESTABLISHED | TCPF_SYN_SENT |
+			         TCPF_CLOSE_WAIT)))
```

Comment:
Linux `tcp_fin()` sets `tcp_set_state(sk, TCP_CLOSE_WAIT)` when we receive FIN from a peer, next the function calls `sk->sk_state_change(sk)`, i.e. our `ss_tcp_state_change()` where we call `ss_close()` now, so our closing work is queued into work queue. `ss_tx_action()` sends all pending data and reaches the closing work. Now the socket in `TCPF_CLOSE_WAIT` state and we do call `__sk_close_locked()`. The change looks correct. We didn't need it before because with previous `ss_linkerror()` we immediately moved from `TCPF_CLOSE_WAIT` to next states like TCP_CLOSE.

Reply:
Yes, it does make sense when it's together with the change in `ss_tcp_state_change()`. Gotcha, and thanks for a clear explanation. This is a good piece of explanation to put somewhere in a comment or, perhaps, just refer to this one in the PR.